### PR TITLE
Use `FIELD_ADDR` for all `ldflda`s

### LINKED ISSUE #77636: The Jit doesn't handle certain uses of field addresses correctly
Reproduction:
```cs
Console.WriteLine(Problem(null));

[MethodImpl(MethodImplOptions.NoInlining)]
static int Problem(StructWithIndex* s)
{
    return *(int*)((nint)(int*)&s->Value | -1);
}

struct StructWithIndex
{
    public int Index;
    public int Value;
}
```
Compile and run.

Expected result: `NullReferenceException` is thrown.

Actual result: `AccessViolationException` is thrown.

Cause: the morphing context used for checking whether an explicit null check is required has fairly non-obvious propagation rules. In particular, it does not get "reset" for arbitrary users, but passed as-is.

### PR DESCRIPTION
**Pull request by SingleAccretion**
I. e. add support for when the underlying address is local.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=95283&view=ms.vss-build-web.run-extensions-tab).

A small number of improvements due to:
1) Minor CSE changes.
2) Less aggressive promotion of implicit byrefs.

There are also some regressions in CoreCLR tests where we stop promoting locals with SIMD fields because `FIELD_ADDR`s don't resolve SIMD type handles. I think it is good TP tradeoff to keep it like that, considering the regressions are test-only and in non-test collections this leads to improvements.

There is also a nice TP win (~`0.3`%).

Fixes #77636.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/lclmorph.cpp [hunk 1]
```diff
@@ -1184,32 +1207,32 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
                     }
 
-                    ClassLayout* layout  = indir->TypeIs(TYP_STRUCT) ? indir->GetLayout(m_compiler) : nullptr;
-                    unsigned     indSize = indir->TypeIs(TYP_STRUCT) ? layout->GetSize() : genTypeSize(indir);
+                    ClassLayout* layout  = node->TypeIs(TYP_STRUCT) ? node->GetLayout(m_compiler) : nullptr;
+                    unsigned     indSize = node->TypeIs(TYP_STRUCT) ? layout->GetSize() : genTypeSize(node);
                     if (indSize > genTypeSize(fieldType))
                     {
-                        // Retargeting this indirection to reference the promoted field would make it
+                        // Retargeting this nodeection to reference the promoted field would make it
```
Comment (SingleAccretion) [inline src/coreclr/jit/lclmorph.cpp:1219]:
```suggestion
                        // Retargeting this indirection to reference the promoted field would make it
```

**Comment by EgorBo** [comment]
@SingleAccretion can you post the link to the diff once it's ready, I see that previous diffs no longer available

**Comment by SingleAccretion** [comment]
Yes, will do.

**Comment by SingleAccretion** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=95283&view=ms.vss-build-web.run-extensions-tab) have been obtained and still look like the ones in the description.

**Review by EgorBo — APPROVED** [review]
Nice TP wins!

**Pull request merged**